### ISSUE TITLE
Release topojson v1.0

### ISSUE THREAD

**Issue by mattijn** [mattijn/topojson#44]
Current version on PyPi is `v1.0rc4`.
Things left to do for release `v1.0`:

- [x] framework of code including options is there
- [x] more docs needed
- [x] travis doesn't work yet

**Comment by mattijn** [comment]
Btw: this was the end of my summer of code. Tomorrow back to work..

**Comment by martinfleis** [comment]
Hi @mattijn, I just came across this repo. Really good work!
In GeoPandas we had some discussion and attempts to have a direct io to Topojson (https://github.com/geopandas/geopandas/pull/645, https://github.com/geopandas/geopandas/pull/610). While it is simple to read through Fiona, we don't have a way of saving to Topojson now.

Instead of the PR mentioned above, I would like to see it done using your package, so I am checking with you to see whether it is a safe (stable) option yet. https://github.com/calvinmetcalf/topojson.py seem to be unmaintained and with a fraction of options.

If you need a hand pushing the last bit, let me know, I might find a moment.

**Comment by mattijn** [comment]
Thanks @martinfleis, for your comment. There are still a few open issues to solve before releasing a 1.0 version. A bit of pushing is definitely helpful. Potential integration with geopandas might be that push! :)

Besides the defined issues, there are two things to be aware of to find out if this repo is useful for integration:
**1. Speed.** 
Computing a topology can be cost-intensive. This whole repo is more or less build around the `shared_paths` function. But there is also the current bottleneck ([here](https://github.com/mattijn/topojson/blob/master/topojson/core/join.py#L155:L165) in code). It is because the `shared_paths` function cannot do broadcasting. I tried using pygeos (its documented in this [issue](https://github.com/pygeos/pygeos/issues/73#issuecomment-552175579)), but the bottleneck seems to be on the GEOS [side](https://geos.osgeo.org/doxygen/classgeos_1_1operation_1_1sharedpaths_1_1SharedPathsOp.html). 

I never compared this repo to the unmaintained https://github.com/calvinmetcalf/topojson.py repo in terms of speed and correctness, but the start principles differs. This repo builds on shapely and numpy, where the other is solely using python.

**2. Toposimplify**
To be honest: I don't really understand the units of the `toposimplify` parameter. The required value is much different for geographic data in degrees or in meters. There is no issue yet, but in the (slowly appearing) documentation page, I mentioned [this](https://mattijn.github.io/topojson/example/settings-tuning.html#toposimplify):

> I don’t really know what the current input value means, but I do know that there is currently NO option to use a %-value (like in mapshaper.org).
> 
> It would be a great contribution if you can make the `toposimplifiy` setting work using percentage as input!

This would also make it possible to have an improved interactive experience, since the exact `toposimplify` value is almost never equal.

So to answer your question, its close to stable, but not yet ready.

**Comment by martinfleis** [comment]
Thank you!

> I don't really understand the units of the toposimplify parameter. 

For Douglas-Pecker algorithm, epsilon is in the same units as geometry, that is why there is so big difference between degrees and meters. I like this animation from Wikipedia, which is clear on what is epsilon.

![image](https://user-images.githubusercontent.com/36797143/79784280-ee250700-8339-11ea-882f-f4c9da367d26.gif)

Epsilon in this case is the half of the width of the blue buffer around the line connecting furthest points. If the point happens to be within the buffer (i.e. its distance from the line connecting furthest points is smaller than epsilon), it gets deleted.

For Visvalingam-Whyatt algorithm, epsilon is different, aerial value, so it will be larger. VW [constructs triangles](https://bost.ocks.org/mike/simplify/) between the points along the line and removes those points, which are associated (are on the top of) with the triangle which area is smaller than epsilon.

I hope it is clear.

In both cases, values are depending on the actual units. It is then complicated to guess default value. That is likely the reason why shapely's simplify does not have default and user has to specify it. That might be good way for `topojson` as well.

> Computing a topology can be cost-intensive

I'll need to explore your code to say anything meaningful on speed during topology creation.

One speedup could be possible in here, if we used vectorized pygeos version of simplify instead of loop through shapely.

https://github.com/mattijn/topojson/blob/238fb29246e0c70967810eaa962e0be86691c5ee/topojson/ops.py#L529-L541

GeoPandas is close to 0.8 release, so I would like include toposimplify (https://github.com/geopandas/geopandas/issues/1387) in 0.9. There's some time to release topojson 1.0 in the meantime.

**Comment by martinfleis** [comment]
> there is currently NO option to use a %-value (like in mapshaper.org)

The % is mapshaper is the number of retained points. I don't think there's any option how to do that using GEOS and `simplification`. You would have to implement the algorithm by yourself and have percentage as a stop point during its iteration.

**Comment by mattijn** [comment]
Thanks for the detailed feedback. Very much appreciated. As you have noticed I started to push the last bits of open issues.

In coming week I'll distill the above mentioned points in a separate issues, and create a new release candidate version (`v1.0rc9)`, based on the current code base.  

In the period after will continue to improve the documentation of the manual and inline docstrings. No new features will be introduced and if no apparent bugs appear the `v1.0` release will be the same as the `v1.0rc9`.

**Comment by martinfleis** [comment]
Just a small followup. After your question I decided to write up the explanation of algorithms here http://martinfleischmann.net/line-simplification-algorithms/

**Comment by mattijn** [comment]
That is great @martinfleis! Really nice explanation of differences between the two epsilons. Love it.

I created https://github.com/mattijn/topojson/issues/73, which provides an example of the time-expensive bottleneck I mentioned above.

**Comment by mattijn** [comment]
@martinfleis, as you might have noticed, I've released version `1.0rc10`.  I said before that there will be no other releases after `rc9`, but after all your feedback + much appreciated suggestions I have made quite some changes that do not yet deserve a full 1.0 release. Any feedback still much welcome!

**Comment by martinfleis** [comment]
@mattijn I've seen, the performance gain is great. What are the remaining bottlenecks now? I did not have a chance to actually see what has changed.

**Comment by mattijn** [comment]
Compared to `rc9` I've worked my way through `Extract`, `Join`, `Cut` and `Dedup` to fix bottlenecks in each of them. In `Hashmap` I've made sure that it works OK with all changes made, but I did not fix or change bottlenecks there. With some timings for larger files (it varies a bit by file) most of the time is now spend in the `_backward_arcs` [function](https://github.com/mattijn/topojson/blob/master/topojson/core/hashmap.py#L182) within the `Hashmap` class. 

If no big issues appear I'll release `1.0` beginning of July and then probably wait till shapely `2.0` has arrived and see what can be done better with the new functionalities in there.

**Comment by martinfleis** [comment]
I'll try to find a moment to play with the master in the following days.

> wait till shapely 2.0 has arrived and see what can be done better with the new functionalities in there.

It will be essentially what is in `pygeos` now, so you can already play with that (I know you already did). If you have pygoes in your environment, you can use geopandas master which uses pygeos geometry under the hood (https://geopandas.readthedocs.io/en/latest/install.html#using-the-optional-pygeos-dependency), so you can work directly with the arrays.

**Comment by mattijn** [comment]
I've checked with geopandas master including pygeos and all still works:
```python
import geopandas as gpd
import topojson as tp

print(f'geopandas: {gpd.__version__}\ntopojson:  {tp.__version__}')
gpd.options.use_pygeos = True

gdf = gpd.read_file(gpd.datasets.get_path("naturalearth_lowres"))
topo = tp.Topology(gdf.query('continent == "Africa"')).toposimplify(4)

topo.to_svg()
topo.to_gdf().head(1)
```

> ```bash
> geopandas: 0.7.0+79.g9f4c995
> topojson:  1.0rc10
> ```
> ![africa](https://user-images.githubusercontent.com/5186265/84927434-22643c00-b0cd-11ea-8401-9ce9acc9a34d.png)
> <div>
> <table border="1" class="dataframe">
>   <thead>
>     <tr style="text-align: right;">
>       <th></th>
>       <th>geometry</th>
>       <th>id</th>
>       <th>pop_est</th>
>       <th>continent</th>
>       <th>name</th>
>       <th>iso_a3</th>
>       <th>gdp_md_est</th>
>     </tr>
>   </thead>
>   <tbody>
>     <tr>
>       <th>0</th>
>       <td>POLYGON ((33.90369 -0.95000, 39.20219 -4.67675...</td>
>       <td>None</td>
>       <td>53950935</td>
>       <td>Africa</td>
>       <td>Tanzania</td>
>       <td>TZA</td>
>       <td>150600.0</td>
>     </tr>
>   </tbody>
> </table>
> </div>
> 

Regarding pygeos related things to investigate:
- Probably its necessary to change a few things related to the `STRtree` 
- It will be interesting to investigate if it is possible to use GEOSRelate (DE-9IM, boundary-boundary) once its landed in pygeos (this could be test with current shapely already).
- Since geometries in pygeos/shapely 2.0 are immutable, maybe can use these and use `numpy.asarray()` representations for a view.

**Comment by martinfleis** [comment]
Indeed, everything should work, there's a compatibility layer in GeoPandas. 

> Probably its necessary to change a few things related to the STRtree

Look at new [`query`](https://github.com/geopandas/geopandas/blob/9f4c9954973f82043a844a011ee7e7b69c639da6/geopandas/sindex.py#L318) and especially [`query_bulk`](https://github.com/geopandas/geopandas/blob/master/geopandas/sindex.py#L364) options. Both are backwards compatible, so they work even without pygeos and are super fast (with pygeos).

**Comment by mattijn** [comment]
Released [v1.0](https://pypi.org/project/topojson/1.0) on PyPi and GitHub tag [v1.0](https://github.com/mattijn/topojson/releases/tag/v1.0) reflects this.
